### PR TITLE
Bluetooth: audio: Expose API methods to change the available contexts

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This exposes API methods to change the available context types.

### CODE REVIEW

**Comment by MariuszSkamra** [comment]
@Thalley I'm wondering how the API for `capabilities.h` should look like ^^ I was about to make is similar to `bt_audio_capability_set_location` however, the user might want to set/clear particular bits of `BT_AUDIO_CONTEXT_TYPE_*`. Thus API method named `*_set` is misleading. Moreover I'm not sure we want to keep the copy of contexts in `capabilities.c` because the user would more likely to keep the data as well.

So I think of rather extending `bt_audio_capability_ops` with getter method.
@Thalley Any suggestions?

**Comment by MariuszSkamra** [comment]
> @Thalley I'm wondering how the API for `capabilities.h` should look like ^^ I was about to make is similar to `bt_audio_capability_set_location` however, the user might want to set/clear particular bits of `BT_AUDIO_CONTEXT_TYPE_*`. Thus API method named `*_set` is misleading. Moreover I'm not sure we want to keep the copy of contexts in `capabilities.c` because the user would more likely to keep the data as well.
> 
> So I think of rather extending `bt_audio_capability_ops` with getter method. @Thalley Any suggestions?

responding myself that `bt_audio_capability_ops` cannot be extended as those are used in the context of `bt_audio_stream`

File: include/zephyr/bluetooth/audio/audio.h [hunk 1]
```diff
@@ -32,93 +32,26 @@ extern "C" {
...
+	BT_AUDIO_CONTEXT_TYPE_GAME = BIT(3),
+	BT_AUDIO_CONTEXT_TYPE_INSTRUCTIONAL = BIT(4),
+	BT_AUDIO_CONTEXT_TYPE_VOICE_ASSISTANTS = BIT(5),
+	BT_AUDIO_CONTEXT_TYPE_LIVE = BIT(6),
+	BT_AUDIO_CONTEXT_TYPE_SOUND_EFFECTS = BIT(7),
+	BT_AUDIO_CONTEXT_TYPE_NOTIFICATIONS = BIT(8),
+	BT_AUDIO_CONTEXT_TYPE_RINGTONE = BIT(9),
+	BT_AUDIO_CONTEXT_TYPE_ALERTS = BIT(10),
+	BT_AUDIO_CONTEXT_TYPE_EMERGENCY_ALARM = BIT(11),
+};
```
Comment (Thalley) [inline include/zephyr/bluetooth/audio/audio.h:53]: Good that it is now an enum :) 

Any reason why you omitted the documentation for each type though?

File: include/zephyr/bluetooth/audio/capabilities.h [hunk 2]
```diff
@@ -230,8 +230,6 @@ struct bt_audio_capability_ops {
 struct bt_audio_capability {
 	/** Capability direction */
 	enum bt_audio_dir dir;
-	/** Supported Audio Contexts */
-	uint16_t context;
```
Comment (Thalley) [inline include/zephyr/bluetooth/audio/capabilities.h:234]: Why was this removed?

**Review by Thalley — CHANGES_REQUESTED** [review]

**Comment by Thalley** [comment]
> > @Thalley I'm wondering how the API for `capabilities.h` should look like ^^ I was about to make is similar to `bt_audio_capability_set_location` however, the user might want to set/clear particular bits of `BT_AUDIO_CONTEXT_TYPE_*`. Thus API method named `*_set` is misleading. Moreover I'm not sure we want to keep the copy of contexts in `capabilities.c` because the user would more likely to keep the data as well.
> > So I think of rather extending `bt_audio_capability_ops` with getter method. @Thalley Any suggestions?
> 
> responding myself that `bt_audio_capability_ops` cannot be extended as those are used in the context of `bt_audio_stream`

It is a very good question. I am not quite sure what to do about the whole `capabilities` part. My original thought was that `capabilities` would be a generic (non-BT) part of Zephyr that would be used to register device capabilities, and then when PACS were read, it would read it from this (and potentially converting the data from a generic storage to a BT PACS specific format). The reason for this is that we can then use the same data/format for not just LE Audio, but also classic audio (if there is any capabilities protocol there), and even perhaps WiFi or USB audio. 

Another thing to consider w.r.t. the changes in this PR, is that PACS is not just for the Unicast Server, but also for the Broadcast Sink (see my changes in https://github.com/zephyrproject-rtos/zephyr/pull/44802).

File: include/zephyr/bluetooth/audio/audio.h (hunk 1 above)
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/audio.h:53]: The names are intuitive and are well documented in assigned numbers

File: include/zephyr/bluetooth/audio/capabilities.h (hunk 2 above)
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/capabilities.h:234]: It was not used - dead code

Comment (Thalley) [reply include/zephyr/bluetooth/audio/capabilities.h:234]: Oh :) 

I assume that it was supposed to hold the supported audio context in PACS which is currently just:
```c
	struct bt_pacs_context context = {
#if defined(CONFIG_BT_PAC_SNK)
		.snk = sys_cpu_to_le16(CONFIG_BT_PACS_SNK_CONTEXT),
#endif
#if defined(CONFIG_BT_PAC_SRC)
		.src = sys_cpu_to_le16(CONFIG_BT_PACS_SRC_CONTEXT),
#endif
	};
```

If the supported context is a compile-time thing, then it's fine to remove :)

File: include/zephyr/bluetooth/audio/audio.h (hunk 1 above)
Comment (Thalley) [reply include/zephyr/bluetooth/audio/audio.h:53]: If we rely on the assigned number documentation, then we should reference that.

Not sure if I agree that the names are enough, as something like `BT_AUDIO_CONTEXT_TYPE_INSTRUCTIONAL` may not be clear, and that it perhaps also isn't clear what the difference between `BT_AUDIO_CONTEXT_TYPE_NOTIFICATIONS` and `BT_AUDIO_CONTEXT_TYPE_SOUND_EFFECTS` is, and how `BT_AUDIO_CONTEXT_TYPE_ALERTS` and `BT_AUDIO_CONTEXT_TYPE_EMERGENCY_ALARM` differ from each other. 

So I suggest to either reference the Generic Audio assigned numbers, or that we keep the documentation.

Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/audio.h:53]: > If we rely on the assigned number documentation, then we should reference that.
> 
> Not sure if I agree that the names are enough, as something like `BT_AUDIO_CONTEXT_TYPE_INSTRUCTIONAL` may not be clear, and that it perhaps also isn't clear what the difference between `BT_AUDIO_CONTEXT_TYPE_NOTIFICATIONS` and `BT_AUDIO_CONTEXT_TYPE_SOUND_EFFECTS` is, and how `BT_AUDIO_CONTEXT_TYPE_ALERTS` and `BT_AUDIO_CONTEXT_TYPE_EMERGENCY_ALARM` differ from each other.
> 
> So I suggest to either reference the Generic Audio assigned numbers, or that we keep the documentation.

```
/** @brief Audio Context Type for Generic Audio
 *
 * These values are defined by the Generic Audio Assigned Numbers, bluetooth.com
 */
enum bt_audio_context {
```
is already there. Is that not enough?

Comment (Thalley) [reply include/zephyr/bluetooth/audio/audio.h:53]: Oh, sorry, must've missed that :)

**Comment by MariuszSkamra** [comment]
> It is a very good question. I am not quite sure what to do about the whole `capabilities` part. My original thought was that `capabilities` would be a generic (non-BT) part of Zephyr that would be used to register device capabilities, and then when PACS were read, it would read it from this (and potentially converting the data from a generic storage to a BT PACS specific format). The reason for this is that we can then use the same data/format for not just LE Audio, but also classic audio (if there is any capabilities protocol there), and even perhaps WiFi or USB audio. 
> 
> 
> 
> Another thing to consider w.r.t. the changes in this PR, is that PACS is not just for the Unicast Server, but also for the Broadcast Sink (see my changes in https://github.com/zephyrproject-rtos/zephyr/pull/44802). 
> 

Having common API for audio devices is nice idea, but should be done via devicetree imo.

Regarding your PR, I'm OK. I'll need to rebase if it'll be merged first

File: subsys/bluetooth/audio/capabilities.c [hunk 3]
```diff
@@ -298,6 +301,28 @@ static int write_location_cb(struct bt_conn *conn, enum bt_audio_dir dir,
...
+	IF_ENABLED(CONFIG_BT_PAC_SRC, (
+		if (dir == BT_AUDIO_DIR_SOURCE) {
+			*contexts = source_available_contexts;
+			return 0;
+		}
+	));
+
+	BT_ERR("Invalid endpoint dir: %u", dir);
+
+	return -EINVAL;
```
Comment (Thalley) [inline subsys/bluetooth/audio/capabilities.c:323]: Should `contexts` be set to 0 in that case? 

Should we really return an error here, as both the available and supported contexts always contains both sink and source contexts?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/pacs.c [hunk 4]
```diff
@@ -608,3 +649,28 @@ int bt_pacs_location_changed(enum bt_audio_dir dir)
...
+	if (err != 0) {
+		BT_WARN("Available Audio Contexts notify failed: %d", err);
+	}
+}
+
+static K_WORK_DELAYABLE_DEFINE(available_contexts_work, available_contexts_notify);
+
+int bt_pacs_available_contexts_changed(void)
+{
+	return k_work_reschedule(&available_contexts_work, PAC_NOTIFY_TIMEOUT);
```
Comment (sjanc) [inline subsys/bluetooth/audio/pacs.c:675]: k_work_reschedule() may return positive value which isn't an error,  is that intended?   bt_pacs_available_contexts_changed() caller seems to assume non-zero is an error

**Review by sjanc — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/capabilities.c (hunk 3 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/capabilities.c:323]: I guess not, as the invalid direction is stack fault. Thus I'll put an assert here instead of debug log.

File: subsys/bluetooth/audio/pacs.c (hunk 4 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/pacs.c:675]: Yes, you have right. Just to explain myself I made this like that, because the timeout here is != `K_NO_WAIT`. In that case `k_work_reschedule` won't return an error according to documentation.
However, I agree with you that it's not best choice as API might change or `PAC_NOTIFY_TIMEOUT` may be changed to `K_NO_WAIT`.

File: subsys/bluetooth/audio/capabilities.c [hunk 5]
```diff
@@ -454,3 +480,45 @@ int bt_audio_capability_set_location(enum bt_audio_dir dir,
...
+	err = set_available_contexts(dir, contexts);
+	if (err) {
+		return err;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_AUDIO_UNICAST_SERVER)) {
+		err = bt_audio_unicast_server_available_contexts_changed();
+		if (err) {
+			BT_DBG("Available contexts weren't notified: %d", err);
+		}
```
[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/capabilities.c:520]: We could ^^

File: subsys/bluetooth/audio/pacs.c (hunk 4 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/pacs.c]: @sjanc You are absolute right. I must overlook this part:
```
1 if delay not K_NO_WAIT and work has been scheduled
```

**Review by sjanc — APPROVED** [review]

File: subsys/bluetooth/audio/capabilities.c [hunk 6]
```diff
@@ -28,6 +28,9 @@
 static sys_slist_t snks;
 static sys_slist_t srcs;
 
+IF_ENABLED(CONFIG_BT_PAC_SNK, (static enum bt_audio_context sink_available_contexts;))
```
Comment (Thalley) [inline subsys/bluetooth/audio/capabilities.c:31]: Was actually not aware of the `IF_ENABLED` macro until now (this will also be the first use of it in the Bluetooth subsystem). 

Is it recommended to use this instead of `#if defined(...)`? 

It doesn't really look like it can be used with with an `else`, and personally I don't think it helps readability at all. I'm not going to block the PR based on this, but just giving my personal input :)

Comment (MariuszSkamra) [reply subsys/bluetooth/audio/capabilities.c:31]: I personally, don't like the code chopped by `#ifdef`'s. With lot of them the code looks messy

Comment (Thalley) [reply subsys/bluetooth/audio/capabilities.c:31]: It's a personal preference :D

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by rymanluk — APPROVED** [review]
